src/lib/client.ts

```diff
@@ -0,0 +1,42 @@
+import { ApiService } from './api/api.service';
+import { AuthResource } from './resources/auth.resource';
+
+export interface PermanentConstructorConfigI {
+  sessionToken: string;
+  mfaToken: string;
+  archiveId: number;
+  apiKey: string;
+}
+
+export class Permanent {
+  private apiKey: string;
+  private sessionToken: string;
+  private mfaToken: string;
+  private archiveId: number;
+
+  private api: ApiService;
+
+  public auth: AuthResource;
+  constructor(config: PermanentConstructorConfigI) {
+    const { sessionToken, mfaToken, archiveId, apiKey } = config;
+
+    this.sessionToken = sessionToken;
+    this.mfaToken = mfaToken;
+    this.archiveId = archiveId;
+    this.apiKey = apiKey;
```

Comment:
While testing, I realized that since this is a boundary between our code and client code, we should probably validate our inputs here; we can't depend on the client code being in TypeScript or respecting our type definitions.

I believe this is purely a developer experience issue, and not a security issue; this code is running in their environment, and does not grant them the ability to do anything they couldn't already do by hitting our API directly.

Calling `new Permanent({})` should give some kind of useful error, like "`sessionToken` is a required parameter" or so. Currently it silently accepts the empty object, and then `isSessionValid()` returns `false` - which, while technically correct, does not make obvious the reason for that.

Reply:
Yes! I actually wanted to throw an error or something if params were missing but the fully strict ruleset wouldn't shut up about always being defined if I tried to check for missing params as they're all required in the interface, and then I forgot to investigate how best to do that with that restriction. Open to suggestions on that!

Reply:
> Yes! I actually wanted to throw an error or something if params were missing but the fully strict ruleset wouldn't shut up about always being defined if I tried to check for missing params as they're all required in the interface, and then I forgot to investigate how best to do that with that restriction. Open to suggestions on that!

@slifty found [Joi](https://joi.dev/), and we're using it in [upload-service](https://github.com/PermanentOrg/upload-service/) - see https://github.com/PermanentOrg/upload-service/pull/11 for our first use of it. It's powerful, and using it is really straightforward and readable - strong recommend!

Reply:
Captured in https://github.com/PermanentOrg/node-sdk/issues/7.